eth/eth_test/common_test.go

```diff
@@ -0,0 +1,205 @@
+	ops           []*testOperator
+	nodeStorage   storage.Storage
+	sim           *simulator.SimulatedBackend
+	boundContract *simcontract.Simcontract
+	auth          *bind.TransactOpts
+	shares        [][]byte
+	execClient    *executionclient.ExecutionClient
+	rpcServer     *rpc.Server
+	httpSrv       *httptest.Server
+	validatorCtrl *mocks.MockController
+}
+
+func setupEnv(
+	t *testing.T,
+	ctx context.Context,
+	testAddresses []*ethcommon.Address,
+	validatorsCount uint64,
+	operatorsCount uint64,
+) (*testEnv, error) {
+	logger := zaptest.NewLogger(t)
+
+	// Create operators RSA keys
+	ops, err := createOperators(operatorsCount, 0)
+	if err != nil {
+		return nil, err
+	}
+
+	validators := make([]*testValidatorData, validatorsCount)
+	shares := make([][]byte, validatorsCount)
+
+	// Create validators, BLS keys, shares
+	for i := 0; i < int(validatorsCount); i++ {
+		validators[i], err = createNewValidator(ops)
+		if err != nil {
+			return nil, err
+		}
+
+		shares[i], err = generateSharesData(validators[i], ops, testAddrAlice, i)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	eh, validatorCtrl, nodeStorage, err := setupEventHandler(t, ctx, logger, ops[0], &testAddrAlice, true)
+
+	if err != nil {
+		return nil, err
+	}
+	if validatorCtrl == nil {
+		return nil, fmt.Errorf("error: validatorCtrl is empty")
+	}
+
+	// Adding testAddresses to the genesis block mostly to specify some balances for them
+	sim := simTestBackend(testAddresses)
+
+	// Create JSON-RPC handler
+	rpcServer, err := sim.Node.RPCHandler()
+	if err != nil {
+		return nil, fmt.Errorf("creatingt rpc server: %w", err)
+	}
+	// Expose handler on a test server with ws open
+	httpSrv := httptest.NewServer(rpcServer.WebsocketHandler([]string{"*"}))
+	addr := "ws:" + strings.TrimPrefix(httpSrv.URL, "http:")
+
+	parsed, err := abi.JSON(strings.NewReader(simcontract.SimcontractMetaData.ABI))
+	if err != nil {
+		httpSrv.Close()
+		return nil, fmt.Errorf("parsing contract abi: %w", err)
+	}
+
+	auth, err := bind.NewKeyedTransactorWithChainID(testKeyAlice, big.NewInt(1337))
+	if err != nil {
+		return nil, err
+	}
+
+	contractAddr, _, _, err := bind.DeployContract(auth, parsed, ethcommon.FromHex(simcontract.SimcontractMetaData.Bin), sim)
+	if err != nil {
+		return nil, fmt.Errorf("deploying contract: %w", err)
+	}
+
+	sim.Commit()
+
+	// Check contract code at the simulated blockchain
+	contractCode, err := sim.CodeAt(ctx, contractAddr, nil)
+	if err != nil {
+		httpSrv.Close()
+		return nil, fmt.Errorf("getting contract code: %w", err)
+	}
+	if len(contractCode) == 0 {
+		return nil, fmt.Errorf("error: contractCode is empty")
+	}
+
+	// Create a client and connect to the simulator
+	client, err := executionclient.New(
+		ctx,
+		addr,
+		contractAddr,
+		executionclient.WithLogger(logger),
+		executionclient.WithFollowDistance(0),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	err = client.Healthy(ctx)
+	if err != nil {
+		// Always returns nil error
+		client.Close()
```

Comment:
Why are these two closed not before each return? Maybe `defer` would suit here better?